src/QirRuntime/lib/QAPI/QAPI.cpp

```diff
@@ -0,0 +1,243 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <assert.h>
+#include <vector>
+
+#include "QAPI.hpp"
+
+#include "IQuantumApi.hpp"
+#include "SimFactory.hpp"
+#include "__quantum__rt.hpp"
+
+using namespace std;
+using namespace Microsoft::Quantum;
+
+static IQuantumApi* GetQuantumExecutor(QuantumExecutionContext* context)
+{
+    return reinterpret_cast<IQuantumApi*>(context->hQuantumApi);
+}
+
+QuantumExecutionContext* QAPI_CreateContext_Toffoli() // NOLINT
+{
+    unique_ptr<IQuantumApi> qapi = CreateToffoliSimulator();
+    SetCurrentQuantumApiForQIR(qapi.get());
+    return new QuantumExecutionContext{reinterpret_cast<uint64_t>(qapi.release())};
+}
+
+QuantumExecutionContext* QAPI_CreateContext_Fullstate() // NOLINT
+{
+    unique_ptr<IQuantumApi> qapi = CreateFullstateSimulator();
+    SetCurrentQuantumApiForQIR(qapi.get());
+    return new QuantumExecutionContext{reinterpret_cast<uint64_t>(qapi.release())};
+}
+
+void QAPI_DestroyContext(QuantumExecutionContext* context) // NOLINT
+{
+    SetCurrentQuantumApiForQIR(nullptr);
+    delete GetQuantumExecutor(context);
+    delete context;
+    context = nullptr;
+}
+
+// Qubit management
+EXPORTQAPI Qubit QAPI_AllocateQubit(QuantumExecutionContext* context) // NOLINT
+{
+    return GetQuantumExecutor(context)->AllocateQubit();
+}
+EXPORTQAPI void QAPI_ReleaseQubit(QuantumExecutionContext* context, Qubit qubit) // NOLINT
+{
+    GetQuantumExecutor(context)->ReleaseQubit(qubit);
+}
+
+// Results
+EXPORTQAPI void QAPI_ReleaseResult(QuantumExecutionContext* context, Result result) // NOLINT
+{
+    GetQuantumExecutor(context)->ReleaseResult(result);
+}
+EXPORTQAPI TernaryBool QAPI_AreEqualResults(QuantumExecutionContext* context, Result r1, Result r2) // NOLINT
+{
+    return GetQuantumExecutor(context)->AreEqualResults(r1, r2);
+}
+EXPORTQAPI ResultValue QAPI_GetResultValue(QuantumExecutionContext* context, Result result) // NOLINT
+{
+    return GetQuantumExecutor(context)->GetResultValue(result);
+}
+EXPORTQAPI Result QAPI_UseZero(QuantumExecutionContext* context) // NOLINT
+{
+    return GetQuantumExecutor(context)->UseZero();
+}
+EXPORTQAPI Result QAPI_UseOne(QuantumExecutionContext* context) // NOLINT
+{
+    return GetQuantumExecutor(context)->UseOne();
+}
+
+// Shortcuts

```

Comment:
All the gate operations are specific for a particular instruction set (whereas allocations and deallocations are rather ubiquitous). It may hence make sense to split them out to reflect that. What is the idea here; What is QAPI to be used for in general? What are the general purpose pieces of the runtime and do you have the target specific pieces split out somewhere?

Reply:
My questions here were also answered yesterday. :) 
To recap (correct me if anything seems wrong): QAPI is to facilitate directly interacting with the runtime infrastructure without going via QIR. Since this is out of scope for now, we don't want to take on the burden of maintaining an additional API at this time. We will hence drop the QAPI from this PR. 
As far as the target specific quantum instruction set goes, we said we will split the IQuantumAPI into two pieces; one for the qubit and result handles that are independent on the quantum instruction set (though may vary depending on the used runtime/backend), and one with all the method that are specific to the instruction set supported by the quantum processor itself (including Message/logging, and asserts).

Reply:
No public C-API anymore